db/db_options_test.cc

```diff
@@ -1033,6 +1034,140 @@ TEST_F(DBOptionsTest, SetFIFOCompactionOptions) {
   ASSERT_EQ(fifo_temp_opt[1].age, 30000);
 }
 
+TEST_F(DBOptionsTest, OffPeakTimes) {
+  Options options;
+  options.create_if_missing = true;
+
+  auto verify_invalid = [&]() {
+    Status s = DBImpl::TEST_ValidateOptions(options);
+    ASSERT_NOK(s);
+    ASSERT_TRUE(s.IsInvalidArgument());
+  };
+
+  auto verify_valid = [&]() {
+    Status s = DBImpl::TEST_ValidateOptions(options);
+    ASSERT_OK(s);
+    ASSERT_FALSE(s.IsInvalidArgument());
+  };
+  std::vector<std::string> invalid_cases = {
+      "06:30-",         "-23:30",  // Both need to be set
+      "06:30-06:30",  // Valid, but start time and end time cannot be the same
+      "12:30 PM-23:30", "12:01AM-11:00PM",  // Invalid format
+      "01:99-22:00",                        // Invalid value for minutes
+      "00:00-24:00",                        // 24:00 is an invalid value
+      "random-value",   "No:No-Hi:Hi",
+  };
+
+  std::vector<std::string> valid_cases = {
+      "",  // Not enabled. Valid case
+      "06:30-11:30",
+      "06:30-23:30",
+      "13:30-14:30",
+      "00:00-23:59",
+      "23:30-01:15",  // From 11:30PM to 1:15AM next day. Valid case
+  };
+
+  for (std::string invalid_case : invalid_cases) {
+    options.daily_offpeak_time_utc = invalid_case;
+    verify_invalid();
+  }
+  for (std::string valid_case : valid_cases) {
+    options.daily_offpeak_time_utc = valid_case;
+    verify_valid();
+  }
+
+  auto verify_is_now_offpeak = [&](int now_utc_hour, int now_utc_minute,
+                                   bool expected) {
+    auto mock_clock = std::make_shared<MockSystemClock>(env_->GetSystemClock());
+    mock_clock->SetCurrentTime(now_utc_hour * 3600 + now_utc_minute * 60);
+    Status s = DBImpl::TEST_ValidateOptions(options);
+    ASSERT_OK(s);
+    auto db_options = MutableDBOptions(options);
+    ASSERT_EQ(expected, db_options.IsNowOffPeak(mock_clock.get()));
+  };
+
+  options.daily_offpeak_time_utc = "";
+  verify_is_now_offpeak(12, 30, false);
+
+  options.daily_offpeak_time_utc = "06:30-11:30";
+  verify_is_now_offpeak(5, 30, false);
+  verify_is_now_offpeak(6, 30, true);
+  verify_is_now_offpeak(10, 30, true);
+  verify_is_now_offpeak(11, 30, true);
+  verify_is_now_offpeak(13, 30, false);
+
+  options.daily_offpeak_time_utc = "23:30-04:30";
+  verify_is_now_offpeak(6, 30, false);
+  verify_is_now_offpeak(23, 30, true);
+  verify_is_now_offpeak(0, 0, true);
+  verify_is_now_offpeak(1, 0, true);
+  verify_is_now_offpeak(4, 30, true);
+  verify_is_now_offpeak(4, 31, false);
+
+  // entire day is offpeak (weird use case, but valid)
```

Comment:
Except one minute, right? ;)

Reply:
Ah you are right. Let me fix the comment and add a special case in the test.